File: backend/routes/quality/accreditation.py
from flask import Blueprint, jsonify, request
from functools import wraps
from supabase_client import get_supabase

# Create blueprint
quality_accreditation_bp = Blueprint('quality_accreditation', __name__)

# Simple CORS decorator
def cors_enabled(f):
    @wraps(f)
    def decorated_function(*args, **kwargs):
        # Handle OPTIONS requests
        if request.method == 'OPTIONS':
            response = jsonify({'status': 'preflight'})
            origin = request.headers.get('Origin', 'http://localhost:3001')
            allowed_origins = ['http://localhost:3000', 'http://localhost:3001', 'http://127.0.0.1:3000', 'http://127.0.0.1:3001']
            if origin in allowed_origins:
                response.headers.add('Access-Control-Allow-Origin', origin)
            response.headers.add('Access-Control-Allow-Credentials', 'true')
            response.headers.add('Access-Control-Allow-Headers', 'Content-Type,Authorization')
            response.headers.add('Access-Control-Allow-Methods', 'GET,POST,PUT,DELETE,OPTIONS')
            return response
        
        # Handle actual requests
        response = f(*args, **kwargs)
        if hasattr(response, 'headers'):
            origin = request.headers.get('Origin', 'http://localhost:3001')
            allowed_origins = ['http://localhost:3000', 'http://localhost:3001', 'http://127.0.0.1:3000', 'http://127.0.0.1:3001']
            if origin in allowed_origins:
                response.headers.add('Access-Control-Allow-Origin', origin)
            response.headers.add('Access-Control-Allow-Credentials', 'true')
        return response
    return decorated_function

@quality_accreditation_bp.route('/accreditation/readiness', methods=['GET', 'OPTIONS'])
@cors_enabled
def get_readiness_score():
    """Get accreditation readiness score from quality_accreditation table"""
    try:
        supabase = get_supabase()

        # Get all records
        result = supabase.table('quality_accreditation').select('*').execute()

        if not result.data:
            return jsonify({
                'success': True,
                'data': {
                    'overall_score': 0,
                    'readiness_level': 'poor',
                    'criteria_scores': {},
                    'department_scores': {}
                }
            })

        # Calculate overall score
        all_scores = [float(record['score']) for record in result.data if record['score']]
        overall_score = sum(all_scores) / len(all_scores) if all_scores else 0

        # Calculate department scores
        department_scores = {}
        for record in result.data:
            dept = record['department']
            score = float(record['score']) if record['score'] else 0

            if dept not in department_scores:
                department_scores[dept] = []
            department_scores[dept].append(score)

        # Average scores per department
        dept_avg_scores = {}
        for dept, scores in department_scores.items():
            dept_avg_scores[dept] = round(sum(scores) / len(scores), 1) if scores else 0

        # Determine readiness level
        if overall_score >= 90:
            readiness_level = 'excellent'
        elif overall_score >= 80:
            readiness_level = 'good'
        elif overall_score >= 70:
            readiness_level = 'average'
        else:
            readiness_level = 'poor'

        # Mock criteria scores (could be enhanced with more specific data)
        criteria_scores = {
            'Curriculum': 88,
            'Faculty': 82,
            'Infrastructure': 90,
            'Research': 78,
            'Student Support': 85,
            'Governance': 87
        }

        readiness_score = {
            'overall_score': round(overall_score, 1),
            'readiness_level': readiness_level,
            'criteria_scores': criteria_scores,
            'department_scores': dept_avg_scores
        }

        return jsonify({
            'success': True,
            'data': readiness_score
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_accreditation_bp.route('/accreditation/analytics', methods=['GET', 'OPTIONS'])
@cors_enabled
def get_accreditation_analytics():
    """Get accreditation analytics"""
    try:
        analytics = {
            'score_trends': [
                {'date': '2024-01', 'score': 78},
                {'date': '2024-02', 'score': 80},
                {'date': '2024-03', 'score': 82},
                {'date': '2024-04', 'score': 81},
                {'date': '2024-05', 'score': 83},
                {'date': '2024-06', 'score': 85}
            ],
            'department_readiness': [
                {'department': 'Computer Science', 'readiness_score': 90},
                {'department': 'Electronics', 'readiness_score': 85},
                {'department': 'Mechanical', 'readiness_score': 82},
                {'department': 'Civil', 'readiness_score': 88}
            ],
            'readiness_distribution': [
                {'level': 'excellent', 'count': 1},
                {'level': 'good', 'count': 2},
                {'level': 'average', 'count': 1},
                {'level': 'poor', 'count': 0}
            ]
        }
        
        return jsonify({
            'success': True,
            'data': analytics
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_accreditation_bp.route('/accreditation', methods=['GET', 'OPTIONS'])
@cors_enabled
def get_accreditation():
    """Get accreditation information from Supabase"""
    try:
        supabase = get_supabase()

        # Get pagination parameters
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', 10))
        offset = (page - 1) * limit

        # Get filter parameters
        search = request.args.get('search', '')
        status_filter = request.args.get('status', '')

        # Build query
        query = supabase.table('quality_accreditation').select('*', count='exact')

        # Apply filters
        if search:
            query = query.or_(f"report_type.ilike.%{search}%,department.ilike.%{search}%")
        if status_filter:
            query = query.eq('status', status_filter)

        # Get total count for pagination
        total_result = supabase.table('quality_accreditation').select('*', count='exact').execute()
        total_count = total_result.count if hasattr(total_result, 'count') else 0

        # Apply pagination
        query = query.range(offset, offset + limit - 1)

        # Execute query
        result = query.execute()

        if not hasattr(result, 'data'):
            return jsonify({
                'success': False,
                'error': 'Failed to fetch data from Supabase'
            }), 500

        # Map database fields to frontend expected fields
        accreditation = []
        for record in result.data:
            mapped_record = {
                'id': record.get('report_id'),
                'body': record.get('report_type', 'NBA'),
                'program': record.get('department', 'General'),
                'status': record.get('status', 'pending'),
                'valid_until': record.get('valid_until', ''),
                'grade': 'A+' if record.get('score') and float(record.get('score', 0)) >= 90 else
                        'A' if record.get('score') and float(record.get('score', 0)) >= 80 else
                        'B+' if record.get('score') and float(record.get('score', 0)) >= 70 else 'B'
            }
            accreditation.append(mapped_record)

        total_pages = (total_count + limit - 1) // limit if total_count > 0 else 1

        return jsonify({
            'success': True,
            'data': accreditation,
            'pagination': {
                'currentPage': page,
                'totalPages': total_pages,
                'totalItems': total_count,
                'limit': limit
            }
        })

    except Exception as e:
        print(f"Error fetching accreditation: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_accreditation_bp.route('/accreditation/reports', methods=['GET', 'OPTIONS'])
@cors_enabled
def get_accreditation_reports():
    """Get accreditation reports with pagination from quality_accreditation table"""
    try:
        supabase = get_supabase()
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', 10))

        # Calculate offset for pagination
        offset = (page - 1) * limit

        # Query the quality_accreditation table with pagination
        query = supabase.table('quality_accreditation').select('*', count='exact')

        # Apply pagination
        result = query.range(offset, offset + limit - 1).execute()

        # Get total count
        total_count = result.count if hasattr(result, 'count') else 0

        # Transform data to match expected format
        reports = []
        for record in result.data:
            reports.append({
                'id': record['report_id'],
                'accreditation_body': record['report_type'],
                'academic_year': record['report_date'][:4] if record['report_date'] else '2024',  # Extract year from date
                'overall_score': float(record['score']) if record['score'] else 0,
                'readiness_level': 'excellent' if record['score'] and float(record['score']) >= 90 else
                                'good' if record['score'] and float(record['score']) >= 80 else
                                'average' if record['score'] and float(record['score']) >= 70 else 'poor',
                'generated_date': record['report_date'],
                'status': 'approved',  # Default status
                'department': record['department'],
                'recommendations': record['recommendations']
            })

        total_pages = (total_count + limit - 1) // limit

        return jsonify({
            'success': True,
            'data': reports,
            'pagination': {
                'page': page,
                'limit': limit,
                'total': total_count,
                'totalPages': total_pages
            }
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_accreditation_bp.route('/accreditation/reports', methods=['POST', 'OPTIONS'])
@cors_enabled
def generate_accreditation_report():
    """Generate accreditation report"""
    try:
        data = request.get_json()
        
        # Mock report generation
        new_report = {
            'id': 999,
            'accreditation_body': data.get('accreditation_body', 'NAAC'),
            'academic_year': data.get('academic_year', '2024'),
            'overall_score': 85,
            'readiness_level': 'good',
            'generated_date': '2024-01-07',
            'status': 'draft'
        }
        
        return jsonify({
            'success': True,
            'data': new_report,
            'message': 'Accreditation report generated successfully'
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

File: backend/controllers/quality/accreditation.js
const { supabase } = require('../../supabase_client');

// Helper function to calculate readiness score
async function calculateReadinessScore() {
  try {
    // Get faculty performance data
    const { data: faculty } = await supabase
      .from('faculty')
      .select('performance_rating, department');

    // Get audit compliance data
    const { data: audits } = await supabase
      .from('audits')
      .select('compliance_score, department');

    // Get grievance resolution data
    const { data: grievances } = await supabase
      .from('grievances')
      .select('status');

    // Get policy compliance data
    const { data: policies } = await supabase
      .from('policies')
      .select('compliance_score');

    // Calculate criteria scores
    const criteriaScores = {
      'Curriculum': calculateFacultyScore(faculty),
      'Teaching-Learning': calculateFacultyScore(faculty) * 0.9,
      'Research': calculateResearchScore(faculty),
      'Infrastructure': 85, // Mock score
      'Student Support': calculateGrievanceScore(grievances),
      'Governance': calculatePolicyScore(policies),
      'Innovative Practices': 80, // Mock score
    };

    // Calculate department scores
    const departmentScores = {};
    const departments = [...new Set(faculty?.map(f => f.department) || [])];
    
    departments.forEach(dept => {
      const deptFaculty = faculty?.filter(f => f.department === dept) || [];
      const deptAudits = audits?.filter(a => a.department === dept) || [];
      
      departmentScores[dept] = Math.round(
        (calculateFacultyScore(deptFaculty) + 
         calculateAuditScore(deptAudits)) / 2
      );
    });

    // Calculate overall score
    const overallScore = Math.round(
      Object.values(criteriaScores).reduce((sum, score) => sum + score, 0) / 
      Object.keys(criteriaScores).length
    );

    return {
      overall_score: overallScore,
      criteria_scores: criteriaScores,
      department_scores: departmentScores
    };
  } catch (error) {
    console.error('Error calculating readiness score:', error);
    return {
      overall_score: 0,
      criteria_scores: {},
      department_scores: {}
    };
  }
}

// Helper functions for score calculations
function calculateFacultyScore(faculty) {
  if (!faculty || faculty.length === 0) return 0;
  const total = faculty.reduce((sum, f) => sum + (f.performance_rating || 0), 0);
  return Math.round(total / faculty.length);
}

function calculateResearchScore(faculty) {
  if (!faculty || faculty.length === 0) return 0;
  const total = faculty.reduce((sum, f) => sum + (f.research_output || 0), 0);
  return Math.min(100, Math.round(total / faculty.length * 10));
}

function calculateGrievanceScore(grievances) {
  if (!grievances || grievances.length === 0) return 100;
  const resolved = grievances.filter(g => g.status === 'resolved').length;
  return Math.round((resolved / grievances.length) * 100);
}

function calculatePolicyScore(policies) {
  if (!policies || policies.length === 0) return 0;
  const total = policies.reduce((sum, p) => sum + (p.compliance_score || 0), 0);
  return Math.round(total / policies.length);
}

function calculateAuditScore(audits) {
  if (!audits || audits.length === 0) return 0;
  const total = audits.reduce((sum, a) => sum + (a.compliance_score || 0), 0);
  return Math.round(total / audits.length);
}

function getReadinessLevel(score) {
  if (score >= 90) return 'excellent';
  if (score >= 75) return 'good';
  if (score >= 60) return 'average';
  return 'poor';
}

function generateRecommendations(readinessScore) {
  const recommendations = [];
  
  Object.entries(readinessScore.criteria_scores).forEach(([criteria, score]) => {
    if (score < 75) {
      recommendations.push(`Improve ${criteria} performance (Current: ${score}%)`);
    }
  });
  
  if (readinessScore.overall_score < 80) {
    recommendations.push('Focus on overall institutional development');
  }
  
  return recommendations;
}

function identifyStrengths(readinessScore) {
  const strengths = [];
  
  Object.entries(readinessScore.criteria_scores).forEach(([criteria, score]) => {
    if (score >= 85) {
      strengths.push(`${criteria} (${score}%)`);
    }
  });
  
  return strengths;
}

function identifyWeaknesses(readinessScore) {
  const weaknesses = [];
  
  Object.entries(readinessScore.criteria_scores).forEach(([criteria, score]) => {
    if (score < 70) {
      weaknesses.push(`${criteria} (${score}%)`);
    }
  });
  
  return weaknesses;
}

// SUB-MODULE 6: ACCREDITATION READINESS REPORTS
exports.getReadinessScore = async (req, res) => {
  try {
    const readinessScore = await calculateReadinessScore();
    
    res.json({
      success: true,
      data: readinessScore
    });
  } catch (error) {
    console.error('Error calculating readiness score:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to calculate readiness score'
    });
  }
};

exports.getAllReports = async (req, res) => {
  try {
    const { page = 1, limit = 10, accreditation_body, status } = req.query;
    
    let query = supabase
      .from('accreditation_reports')
      .select('*', { count: 'exact' })
      .order('generated_date', { ascending: false });

    if (accreditation_body) query = query.eq('accreditation_body', accreditation_body);
    if (status) query = query.eq('status', status);

    const from = (page - 1) * limit;
    const to = from + limit - 1;

    const { data, error, count } = await query.range(from, to);

    if (error) throw error;

    res.json({
      success: true,
      data: data || [],
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total: count || 0,
        totalPages: Math.ceil((count || 0) / limit)
      }
    });
  } catch (error) {
    console.error('Error fetching accreditation reports:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch accreditation reports'
    });
  }
};

exports.generateReport = async (req, res) => {
  try {
    const { accreditation_body, academic_year } = req.body;
    
    // Calculate readiness score and generate report
    const readinessScore = await calculateReadinessScore();
    
    const reportData = {
      accreditation_body,
      academic_year,
      overall_score: readinessScore.overall_score,
      criteria_scores: readinessScore.criteria_scores,
      department_scores: readinessScore.department_scores,
      readiness_level: getReadinessLevel(readinessScore.overall_score),
      recommendations: generateRecommendations(readinessScore),
      strengths: identifyStrengths(readinessScore),
      weaknesses: identifyWeaknesses(readinessScore),
      generated_date: new Date().toISOString(),
      status: 'draft'
    };

    const { data, error } = await supabase
      .from('accreditation_reports')
      .insert([reportData])
      .select();

    if (error) throw error;

    res.status(201).json({
      success: true,
      data: data[0],
      message: 'Accreditation report generated successfully'
    });
  } catch (error) {
    console.error('Error generating accreditation report:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to generate accreditation report'
    });
  }
};

exports.getAccreditationAnalytics = async (req, res) => {
  try {
    // Get historical reports for trend analysis
    const { data: reports } = await supabase
      .from('accreditation_reports')
      .select('overall_score, generated_date, accreditation_body')
      .order('generated_date', { ascending: true });

    // Score trends
    const scoreTrends = reports ? 
      reports.map(report => ({
        date: new Date(report.generated_date).toLocaleDateString(),
        score: report.overall_score,
        body: report.accreditation_body
      })) : [];

    // Department-wise readiness
    const { data: faculty } = await supabase
      .from('faculty')
      .select('department, performance_rating');

    const departmentReadiness = faculty ? 
      faculty.reduce((acc, faculty) => {
        const dept = acc.find(item => item.department === faculty.department);
        if (dept) {
          dept.total += faculty.performance_rating;
          dept.count += 1;
        } else {
          acc.push({
            department: faculty.department,
            total: faculty.performance_rating,
            count: 1
          });
        }
        return acc;
      }, []).map(item => ({
        department: item.department,
        readiness_score: Math.round(item.total / item.count)
      })) : [];

    // Readiness distribution
    const readinessDistribution = [
      { level: 'Excellent', count: 0, range: '90-100' },
      { level: 'Good', count: 0, range: '75-89' },
      { level: 'Average', count: 0, range: '60-74' },
      { level: 'Poor', count: 0, range: '0-59' }
    ];

    if (reports) {
      reports.forEach(report => {
        if (report.overall_score >= 90) readinessDistribution[0].count++;
        else if (report.overall_score >= 75) readinessDistribution[1].count++;
        else if (report.overall_score >= 60) readinessDistribution[2].count++;
        else readinessDistribution[3].count++;
      });
    }

    const analytics = {
      score_trends: scoreTrends,
      department_readiness: departmentReadiness,
      readiness_distribution: readinessDistribution
    };

    res.json({
      success: true,
      data: analytics
    });
  } catch (error) {
    console.error('Error fetching accreditation analytics:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch accreditation analytics'
    });
  }
};
